esp/esp/utils/query_builder.py

```diff
@@ -0,0 +1,333 @@
+            'filters': {f.name: f.spec() for f in self.filters},
+        }
+
+    def as_queryset(self, value):
+        """Given data returned by the client, return a QuerySet for the query.
+
+        The data returned will be in the format specified in query-builder.jsx.
+        """
+        if value['filter'] in ['and', 'or']:
+            if value['filter'] == 'and':
+                op = operator.and_
+            else:
+                op = operator.or_
+            combined = reduce(op, map(self.as_queryset, value['values']))
+            if value['negated']:
+                return self.base.exclude(pk__in=combined)
+            else:
+                return combined
+        elif value['filter'] in self.filter_dict:
+            filter_obj = self.filter_dict[value['filter']]
+            filter_q = filter_obj.as_q(value['values'])
+            if value['negated'] ^ filter_obj.inverted:
+                # Due to https://code.djangoproject.com/ticket/14645, we have
+                # to write this query a little weirdly.
+                return self.base.exclude(id__in=self.base.filter(filter_q))
+            else:
+                return self.base.filter(filter_q)
+        else:
+            raise ESPError('Invalid filter %s' % value.get('filter'))
+
+    def as_english(self, value, root=True):
+        """Given data returned by the client, return an English description.
+
+        The data returned will be in the format specified in query-builder.jsx.
+        """
+        if root:
+            base = '%s with ' % self.english_name
+        else:
+            base = ''
+        if value['filter'] in ['and', 'or']:
+            if value['negated'] and value['filter'] == 'or':
+                join_word = ' nor '
+                intro_word = ' neither'
+            else:
+                join_word = ' %s ' % value['filter']
+                if value['negated']:
+                    intro_word = ' none of'
+                else:
+                    intro_word = ''
+            return base + intro_word + join_word.join(
+                ['(' + self.as_english(v, root=False) + ')'
+                 for v in value['values']])
+        elif value['filter'] in self.filter_dict:
+            filter_obj = self.filter_dict[value['filter']]
+            if value['negated']:
+                base = base + 'not '
+            return base + filter_obj.as_english(value['values'])
+        else:
+            raise ESPError('Invalid filter %s' % value.get('filter'))
+
+    def render(self):
+        """Render the QueryBuilder into HTML.
+
+        query-builder.jsx will need to be included in the page separately.
+        """
+        context = {
+            # use a uid in case we ever want to have multiple QBs on the same
+            # page.
+            'uid': random.randrange(0, 1 << 30),
+            'json_spec': json.dumps(self.spec()),
+        }
+        return render_to_string("utils/query_builder.html", context)
```

Comment:
I think a template include might be more idiomatic, but I don't like it, because it requires more care in passing pieces around.  It might also be possible to do it as an inclusion tag of some sort, which would be nice in that it could mark the output as safe, so you wouldn't have to do `{{QB.render|safe}}`.  I can take a look at doing that.